addons/portal/controllers/mail.py

```diff
@@ -15,7 +15,10 @@
 def _check_special_access(res_model, res_id, token='', _hash='', pid=False):
     record = request.env[res_model].browse(res_id).sudo()
     if _hash and pid:  # Signed Token Case: hash implies token is signed by partner pid
-        return consteq(_hash, record._sign_token(pid))
+        return (
+            consteq(_hash, record._sign_token(pid)) or
+            consteq(_hash, record._get_parent_sign_token(pid))
+        )
```

Comment:
In the case, `consteq(_hash, record._sign_token(pid))` is falsy with `res_model='project.project'` and `res_id=1` and random `pid` as parameter, your changes will return a traceback because we cannot compare a str with bool value. To avoid that issue, we have to check the value of `record._get_parent_sign_token` before comparing. Something like that:
```suggestion
        can_access = consteq(_hash, record._sign_token(pid))
        if not can_access:
            parent_sign_token = record._get_parent_sign_token(pid)
            can_access = parent_sign_token and consteq(_hash, parent_sign_token)
        return can_access
```